pkg/kubelet/images/image_manager_test.go

```diff
@@ -209,7 +239,9 @@ func TestParallelPuller(t *testing.T) {
 				fakeRuntime.CalledFunctions = nil
 				fakeClock.Step(time.Second)
 				_, _, err := puller.EnsureImageExists(ctx, pod, container, nil, nil)
-				fakeRuntime.AssertCalls(expected.calls)
+				if len(expected.calls) > 0 {
```

Comment:
@kannon92 This looks like we're ignoring calls if we're not expecting them. Why?

Reply:
I really don't understand how we do mocking of functions here.  

I found that all the calls use `GetImageRef` and I notice that is a function in the ImageService.  I'm not sure if I have to refactor this and add functions to the TestRuntime in order to properly mock them.  

So I found that if I don't put an expect call I can still assert against errors that aren't present in the FakeRuntime.  
```
func (f *FakeRuntime) GetImageRef(_ context.Context, image kubecontainer.ImageSpec) (string, error) {
	f.Lock()
	defer f.Unlock()

	f.CalledFunctions = append(f.CalledFunctions, "GetImageRef")
	for _, i := range f.ImageList {
		if i.ID == image.Image {
			return i.ID, nil
		}
	}
	return "", f.InspectErr
}
```

Reply:
it looks like a bug in the FakeRuntime init code. You can try to apply this patch to your PR branch. It should fix panics and reveal test failures:
```
diff --git a/pkg/kubelet/images/image_manager_test.go b/pkg/kubelet/images/image_manager_test.go
index 64f3e1d9738..faf3062d3ec 100644
--- a/pkg/kubelet/images/image_manager_test.go
+++ b/pkg/kubelet/images/image_manager_test.go
@@ -196,7 +196,7 @@ func (m *mockPodPullingTimeRecorder) RecordImageStartedPulling(podUID types.UID)
 
 func (m *mockPodPullingTimeRecorder) RecordImageFinishedPulling(podUID types.UID) {}
 
-func pullerTestEnv(c pullerTestCase, serialized bool) (puller ImageManager, fakeClock *testingclock.FakeClock, fakeRuntime *ctest.FakeRuntime, container *v1.Container) {
+func pullerTestEnv(t *testing.T, c pullerTestCase, serialized bool) (puller ImageManager, fakeClock *testingclock.FakeClock, fakeRuntime *ctest.FakeRuntime, container *v1.Container) {
        container = &v1.Container{
                Name:            "container_name",
                Image:           c.containerImage,
@@ -207,7 +207,7 @@ func pullerTestEnv(c pullerTestCase, serialized bool) (puller ImageManager, fake
        fakeClock = testingclock.NewFakeClock(time.Now())
        backOff.Clock = fakeClock
 
-       fakeRuntime = &ctest.FakeRuntime{}
+       fakeRuntime = &ctest.FakeRuntime{T: t}
        fakeRecorder := &record.FakeRecorder{}
 
        fakeRuntime.ImageList = []Image{{ID: "present_image:latest"}}
@@ -231,17 +231,15 @@ func TestParallelPuller(t *testing.T) {
 
        useSerializedEnv := false
        for _, c := range cases {
-               puller, fakeClock, fakeRuntime, container := pullerTestEnv(c, useSerializedEnv)
+               puller, fakeClock, fakeRuntime, container := pullerTestEnv(t, c, useSerializedEnv)
 
                t.Run(c.testName, func(t *testing.T) {
                        ctx := context.Background()
                        for _, expected := range c.expected {
-                               fakeRuntime.CalledFunctions = nil
+                               fakeRuntime.ClearCalls()
                                fakeClock.Step(time.Second)
                                _, _, err := puller.EnsureImageExists(ctx, pod, container, nil, nil)
-                               if len(expected.calls) > 0 {
-                                       fakeRuntime.AssertCalls(expected.calls)
-                               }
+                               fakeRuntime.AssertCalls(expected.calls)
                                assert.Equal(t, expected.err, err)
                        }
                })
@@ -261,7 +259,7 @@ func TestSerializedPuller(t *testing.T) {
 
        useSerializedEnv := true
        for _, c := range cases {
-               puller, fakeClock, fakeRuntime, container := pullerTestEnv(c, useSerializedEnv)
+               puller, fakeClock, fakeRuntime, container := pullerTestEnv(t, c, useSerializedEnv)
 
                t.Run(c.testName, func(t *testing.T) {
                        ctx := context.Background()
@@ -322,7 +320,7 @@ func TestPullAndListImageWithPodAnnotations(t *testing.T) {
                }}
 
        useSerializedEnv := true
-       puller, fakeClock, fakeRuntime, container := pullerTestEnv(c, useSerializedEnv)
+       puller, fakeClock, fakeRuntime, container := pullerTestEnv(t, c, useSerializedEnv)
        fakeRuntime.CalledFunctions = nil
        fakeRuntime.ImageList = []Image{}
        fakeClock.Step(time.Second)
```

Reply:
Applied.

Reply:
I decided to move some of this into a separate PR: https://github.com/kubernetes/kubernetes/pull/116231

This was mostly so we can discuss the feature separate from some test fixes and additions.